File: LinearRegression/LinearRegression.R
#read the data from csv
Data=read.csv('/Users/Dongzhe/Desktop/stat 202a/state_x77.csv')#already a data frame
str(Data)
#the data has already trimmed to deal with 

summary(Data)
#take some of the attributes from the data
X=Data[c("Population","Income","Illiteracy","Murder","HS.Grad","Frost","Density")]
Y=Data["Life.Exp"]
row = nrow(X)
A = cbind(rep(1, row), X, Y)



mylm_qr<-function(Z){
  n = nrow(Z)
  m = ncol(Z)
  R = Z
  Q = diag(n)
  
  for (k in 1:(m-1))
  {
    x = matrix(rep(0, n), nrow = n)
    x[k:n, 1] = R[k:n, k]
    g = sqrt(sum(x^2))
    v = x
    v[k] = x[k] + sign(x[k,1])*g
    s = sqrt(sum(v^2))
    if (s != 0)
    {
      u = v / s
      R = R - 2 * u %*% t(u) %*% as.matrix(R)
      Q = Q - 2 * u %*% t(u) %*% as.matrix(Q)
    }
  }
  result=list(t(Q), R)
  names(result)=c("Q", "R")
  R1 = R[1:(m-1), 1:(m-1)]
  Y1 = R[1:(m-1), m]
  #return the coefficients only
  beta = solve(R1, Y1)
  beta
}


model1 = lm(Life.Exp ~ Population+Income+Illiteracy+Murder+HS.Grad+Frost+Density,data=Data)
coef(model1)
summary(model1)
mylm_qr(A)
#compare the result between the linear model and the qr result. The function is correct!

#(Intercept)    Population        Income    Illiteracy        Murder       HS.Grad
#7.094143e+01  6.248966e-05  1.484507e-04  1.451796e-01 -3.319248e-01  3.746457e-02
#Frost       Density
#-5.532853e-03 -7.995441e-04
#rep(1, row)    Population        Income    Illiteracy        Murder       HS.Grad 
#7.094143e+01  6.248966e-05  1.484507e-04  1.451796e-01 -3.319248e-01  3.746457e-02 
#Frost       Density 
#-5.532853e-03 -7.995441e-04 


mylm_sweep<-function(A, m)
{
  n <- dim(A)[1]
  
  for (k in 1:m) 
  {
    for (i in 1:n)     
      for (j in 1:n)   
        if (i!=k  & j!=k)     
          A[i,j] <- A[i,j] - A[i,k]*A[k,j]/A[k,k]    
    
    for (i in 1:n) 
      if (i!=k) 
        A[i,k] <- A[i,k]/A[k,k]  
    
    for (j in 1:n) 
      if (j!=k) 
        A[k,j] <- A[k,j]/A[k,k]  
    
    A[k,k] <- - 1/A[k,k] 
  }
  beta = A[1:m, m+1]
  beta
}

X=Data[c("Population","Income","Illiteracy","Murder","HS.Grad","Frost","Density")]
Y=Data["Life.Exp"]
col = ncol(X)
A = cbind(rep(1, row), X, Y)
Z=t(A)%*%as.matrix(A)
#this time input matrix is the a'a

mylm_sweep(Z,col+1)
summary(model1)
#The function is correct

#rep(1, row)    Population        Income    Illiteracy        Murder       HS.Grad 
#7.094143e+01  6.248966e-05  1.484507e-04  1.451796e-01 -3.319248e-01  3.746457e-02 
#Frost       Density 
#-5.532853e-03 -7.995441e-04 

File: 2.LinearRegression/LinearRegression.py

# coding: utf-8


#Dongzheli 104434089

import numpy as np
from scipy import linalg
import pandas as pd
import warnings
warnings.simplefilter('ignore', DeprecationWarning)

data = pd.read_csv('/Users/Dongzhe/Desktop/stat 202a/state_x77.csv')
data.head()

X=data.get(['Population', 'Income', 'Illiteracy','Murder','HS.Grad','Frost','Density' ])
X.head()


Y=data.get(['Life.Exp'])
Y.head()


nx, mx = X.shape
Z = np.hstack((np.ones(nx).reshape((nx, 1)),X,Y))
Z.shape


def mylm_qr(A):
    n, m = A.shape
    R = A.copy()
    Q = np.eye(n)

    for k in range(m-1):
        x = np.zeros((n, 1))
        x[k:, 0] = R[k:, k]
        v = x
        v[k] = x[k] + np.sign(x[k,0]) * np.linalg.norm(x)

        s = np.linalg.norm(v)
        if s != 0:
            u = v / s
            R -= 2 * np.dot(u, np.dot(u.T, R))
            Q -= 2 * np.dot(u, np.dot(u.T, Q))
    Q = Q.T
    R1 = R[:m-1, :m-1]
    Y1 = R[:m-1, m-1]
    beta = np.linalg.solve(R1, Y1)
    return beta

print beta


def mylm_sweep(B, m):
    A = np.copy(B)
    n, c = A.shape
    for k in range(m):
        for i in range(n):
            for j in range(n):
                if i != k and j != k:
                    A[i,j] = A[i,j]- A[i,k]*A[k,j]/A[k,k]

        for i in range(n):
            if i != k:
                A[i,k] = A[i,k]/A[k,k]

        for j in range(n):
            if j != k:
                A[k,j] = A[k,j]/A[k,k]

        A[k,k] = -1/A[k,k]
    beta=A[0:m,m]
    return beta

def eigen_qr(A):
    T = 1000
    A_copy = A.copy()
    r, c = A_copy.shape

    V = np.random.random_sample((r, r))

    for i in range(T):
        Q, _ = qr(V)
        V = np.dot(A_copy, Q)

    Q, R = qr(V)

    return R.diagonal(), Q

Zin=np.dot(Z.T,Z)
mylm_sweep(Zin, mx+1)




